### ISSUE TITLE
Swatches are not crossed out if salable quantity is 0

### ISSUE THREAD

**Issue by chequille** [magento/magento2#33799]
### Preconditions (*)

1. Magento 2,4,3
2.

### Steps to reproduce (*)

1. Make a configurable product with f.e. 3 options
2. ![grafik](https://user-images.githubusercontent.com/15807794/129438093-70ab92fc-28e8-4f12-92bd-21e47d24abad.png)
3. Place an order for this product with one of the options so salable quantity is 0 after the order is placed. In the sample of the screenshot it is the product with option 0.1l and qty 3
4. After this order, in backend salable qty is 0 for the 0.1l product
5. ![grafik](https://user-images.githubusercontent.com/15807794/129438277-f8317351-5597-4aab-906e-365e7f4233d4.png)
6. But in frontend, the swatch for 0.1l is still selectable and not crossed out as it should.
7. ![grafik](https://user-images.githubusercontent.com/15807794/129438326-0821c897-4675-43dd-805c-01bf5e3ac1fa.png)

### Expected result (*)

1. The swatch option 0.1l should be crossed and not selectable anymore as it worked in Mangeot 2.3.x
2.

### Actual result (*)

1. The swatch option 0.1l is not crossed and still seleactable.
2.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by chequille** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by chequille** [comment]
Reproduced on 2.4.develop instance.
Ordered 100 Chaz Kangeroo Hoodie XS black but in frontend XS is still selectable
![grafik](https://user-images.githubusercontent.com/15807794/129451018-1a663b94-e726-4826-a53d-40e51af9c238.png)

If I select it and add it to cart, than it is crossed and the error message comme up that I have to selet an option.
![grafik](https://user-images.githubusercontent.com/15807794/129451094-585cdded-8fac-4aae-abc7-277e0f79ea78.png)
But the crossed XS must be there as soon as the product page is launched.

Please correct this as fast as possible.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @krishprakash 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-43089` were created

**Issue Available**: @krishprakash, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by chequille** [comment]
HI Guys,
saw that you put this on P2
I would like to have this at least P1

I am paying since more than 9 month now for a VServer to be able to switch to 2.4.x and cannot because of this bug.
Please consider to put this on highest priority as it is a mandatory feature and many other installation wil have this problem as well.
I would like to get a patch as soon as you found the problem. I definetely cannot wait another 3 or more months for the next release.

BR
Chequille

**Comment by Kannakiraj123** [comment]
@magento assign this to me

**Comment by Kannakiraj123** [comment]
@magento I am working on this

**Comment by Kannakiraj123** [comment]
@magento give me 2.4-develop instance

**Comment by dphilipps** [comment]
I too have been waiting patiently for this fix and very disappointed to find such a crucial feature is still not working properly.  What is the point of maintaining a salable_quantity if the store still shows unsalable products as in stock?!!

I've spent many hours going around in circles on this issue but I have established the following facts:

- When all indexers are set to 'Scheduled' and cron is NOT running, the store correctly crosses out unsaleable products when salable_qty becomes 0 - it works.
- After a product is completely sold out and correctly crossed out, invalidating the indexes via 'Index Management' in admin will **cause the swatch to revert** to showing the product as salable again (i.e. no longer crossed out)
- Likewise, running /php bin/magento indexer:reindex from the command line will revert all unsalable products so they appear in stock on the front end of the store.  Running cron with scheduled indexes normally has a similar effect.
- After purchasing all salable stock for a product, the stock_status column is automatically set to 0 in the cataloginventory_stock_status table.  The view inventory_stock_1 reflects this accordingly in the 'is_salable' column.
- Immediately after reindex, invalidation, or running cron, this value reverts to 1 for any test cases that were working previously.

Sporadically using a magento deployment instance the feature does work correctly and crosses out options after purchasing 100 products.  Often it fails first time then works on subsequent attempts - this is with the Indexers set to 'Update on Save' as per the default.  By switching indexers to Scheduled as above (I believe cron is not running on the Magento instances?) it works correctly every time for me.

The bottom line is it appears at least one of the major issues is with the re-indexing of the salable status.  I believe it may be reverting the salable status of a product immediately after it is sold out before you can navigate back to the product page to check it is working.  I hope this is of some help to fixing and verifying this feature as working.

**Comment by vseager** [comment]
This issue is still reproducible on 2.4.3 for both configurable and simple products.

**Comment by dphilipps** [comment]
Hi @chequille 
Have you checked that consumers are running, specifically **inventory.reservations.updateSalabilityStatus**

This consumer is responsible for changing the salability status after a product is completely sold out.  And because it runs as an asynchronous process, there is sometimes a delay in the sold out product being crossed out on the front-end.  Without the consumer running, it does not become crossed out at all.

There is a different issue I have noticed where reindexing will revert the salability status of products so they are no longer crossed out.  However, the bug has been closed so I'll try and get it reopened: https://github.com/magento/magento2/issues/34127

**Comment by chequille** [comment]
Hi @dphilipps ,
must be honest, I do not know how to check. As well, as I do not need it currently, I don't have cronjobs running.
But I know, that this did all work in Magento 2.3.x wihtout problem.

My solution currently is to make the changes in Magento/InventoryIndexer/Model/IsProductSalable.php as describe in https://github.com/magento/magento2/pull/33885

Fact as I see it is, that as I said before isSaleable() and isSalable() methods do not car about salable quantity. They just look at stock quantity and return true if it is not 0.

The change in the class IsProductSalable changes this and it works for me at the moment. THe only thing I did have to do is to make my product view site and the calatolg view site not cachable.

WIth this changes, "add to cart" button and as well the crossed options of a configurable product are correct as they should.

As for the moment, I can live with the fact of not caching the page, but there should be a fix in the future in any case, as it is a shame that this issue makes that much problems.

My opinion!

Regards,
Chequille

**Comment by engcom-Bravo** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Bravo** [comment]
Hi @chequille,


Thank you for reporting and collaboration. Verified the issue on Magento 2.4-develop branch but could not able to reproduce the issue. Kindly refer the screenshots.


1.Created a Configurable product with 3 options.

<img width="938" alt="000" src="https://user-images.githubusercontent.com/51680745/188630174-d89cf4b7-5b97-4702-ad99-8a85d3507843.png">




2.Order placed with qty 3(color-gray).


<img width="890" alt="001" src="https://user-images.githubusercontent.com/51680745/188630689-cca19f3a-edad-4deb-8ba3-26496f14c0c2.png">



3.After placing the order the quantity is 0.



<img width="931" alt="002" src="https://user-images.githubusercontent.com/51680745/188631150-fd8691b8-4ad4-4f50-aefa-39e7e8d2ad76.png">




4.In frontend we are not able to see that grey color swatch.



<img width="897" alt="003" src="https://user-images.githubusercontent.com/51680745/188631392-a8a738a6-1b62-41df-aee3-cf66a395204d.png">




Let us Know we miss anything.


Thanks.

**Comment by mysil2** [comment]
@engcom-Bravo This needs to be run with MSI. Quantity should still stay at 3 when order is in pending/processing state, but Salable quantity will go from 3 to 0, and the bug will still show the swatch. When order is shipped/complete, then quantity is set to 0 and swatch is not shown.

**Comment by chequille** [comment]
I agree to mysil. This happen when using MSI which is standard as I know. Am I wrong?

**Comment by engcom-Bravo** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Bravo** [comment]
Hi @chequille,


Thank you for reporting and collaboration.


Verified the issue on Magento 2.4-develop branch we are able to reproduce the issue. Kindly refer the screenshots.


1.Created a Configurable Product with 3 Customizable options.

<img width="940" alt="a" src="https://user-images.githubusercontent.com/51680745/188872057-7a1c5965-89c1-446d-9a65-02f8c4de4ec3.png">



2.Placing an order for this product with one of the options(Green Color).


<img width="919" alt="b" src="https://user-images.githubusercontent.com/51680745/188872530-a61974ef-7c6d-4851-8398-6c1e16dc40c9.png">



3.After Placing an order the salable qty is o in backend.


<img width="939" alt="c" src="https://user-images.githubusercontent.com/51680745/188872765-86c6cfa6-dd13-4e21-a60b-51bbf7eb5a24.png">



4.But in frontend the swatch(Green Color) is still selectable and not crossed out.


<img width="934" alt="e" src="https://user-images.githubusercontent.com/51680745/188873153-3a58466b-9b5a-42ab-8e1f-b621c11840ba.png">



Hence Confirming the issue.


Thanks.

**Comment by leonhelmus** [comment]
@chequille For now it seems this patch resolves my issues:

```
diff --git a/vendor/magento/module-configurable-product/Model/ConfigurableAttributeData.php b/vendor/magento/module-configurable-product/Model/ConfigurableAttributeData.php
--- a/vendor/magento/module-configurable-product/Model/ConfigurableAttributeData.php	
+++ b/vendor/magento/module-configurable-product/Model/ConfigurableAttributeData.php	(date 1662571233593)
@@ -58,12 +58,21 @@
         $attributeOptionsData = [];
         foreach ($attribute->getOptions() as $attributeOption) {
             $optionId = $attributeOption['value_index'];
+
+            if (isset($config['salable'])) {
+                $products = isset($config['salable'][$attribute->getAttributeId()][$optionId])
+                    ? $config['salable'][$attribute->getAttributeId()][$optionId]
+                    : [];
+            } else {
+                $products = isset($config[$attribute->getAttributeId()][$optionId])
+                    ? $config[$attribute->getAttributeId()][$optionId]
+                    : [];
+            }
+
             $attributeOptionsData[] = [
                 'id' => $optionId,
                 'label' => $attributeOption['label'],
-                'products' => isset($config[$attribute->getAttributeId()][$optionId])
-                    ? $config[$attribute->getAttributeId()][$optionId]
-                    : [],
+                'products' => $products
             ];
         }
         return $attributeOptionsData;
```

**Comment by chequille** [comment]
Hi Guys,
first of all wondering that now this issue is investigated again. But I guess it is because some other people facing the same problem still with 2.4.4 version.
Currently, I do not touch my live system, as it is working there. I could solve this with the help from Kannakiraj123, SHe introduced a change in InventoryIndexer/IsSalable.php class which solved the problem. There was as well another change which I did not find currently. It was as well something with attributeOptions.
It seems, that the change in InvntoryIndexer was never applied to the core.

Maybe the change above in ConfigurableAttributeData.php will solve the problem as well, do not know yet, but I will try this with my test system.

Coming back to ou with the results as soon as I had the time to do it.

Best regards,
Chequille

**Comment by engcom-Dash** [comment]
@magento give me 2.4-develop instance

**Comment by RajasekarPanneerselvam** [comment]
@magento give me 2.4-develop instance

**Comment by pmathbliss** [comment]
I found a similar issue where the cataloginventory_stock_status is not updating when reindexing. 
After deleting the cataloginventory_stock_status rows for the product, the indexer reindexed properly.

**Comment by Kannakiraj123** [comment]
hi @chequille , still Is this issue reproducing?

**Comment by Kannakiraj123** [comment]
@magento I am working on this

**Comment by chequille** [comment]
Hi Everybody,
must say, that I am currently still on Magento 2.4.3, as I could not update to current version because one of my main extension not working on 2.4.4 or higher.
Therefore, I am using the commit 372ccaa to get swtaches crossed if sailable quantity is 0 and it is working.
Only one thing I have to do: all catalgog and product pages are not cashed. If I would enable cash in the layout file, it would not work till I run clear cash. After clearing cash, swatches are crossed as the should according to the sailabe quantity.
As my server is fast enough, I can live with this.

This is all I can say regarding this issue at the moment.
Best regards,
Chequille

**Comment by Kannakiraj123** [comment]
hi @chequille @engcom-Bravo 


Verified the issue on Magento 2.4-develop branch we are not able to reproduce the issue. Kindly refer the screenshots.but after placing order, when i go to product view page swatches are not crossed out at immediately .we have to clear cache or try to add to cart with options swatches are crossed out.

1.Created a Configurable Product with 3 Customizable options and place the order with one options(s size)

<img width="1739" alt="Screenshot 2023-01-23 at 8 16 19 AM" src="https://user-images.githubusercontent.com/88073031/213958260-7b0451db-0ba6-4332-ae20-8bd7e2d4a148.png">


<img width="1609" alt="Screenshot 2023-01-23 at 8 16 38 AM" src="https://user-images.githubusercontent.com/88073031/213958280-69238d39-d7d8-4262-b488-5b766f173469.png">

**Comment by chequille** [comment]
Hi,
that is exactly what is going wrong. You have to clear cache before the swatch is crossed out.

I am still on 2.4.3 with my patch to get it working.

Best regards,
Chequille

**Comment by engcom-Bravo** [comment]
@Kannakiraj123 Thanks for your contribution.

Hi @chequille,

We have verified the issue in Latest 2.4-develop instance and the issue is no more reproducible.Kindly refer the screenshots.

<img width="1340" alt="Screenshot 2024-07-31 at 11 00 26" src="https://github.com/user-attachments/assets/9f79cf4a-fd6e-431f-8fd6-bb1efeea98d9">



Swatches are crossed out if the salable quantity is 0.

Hence we are closing this issue.

Thanks.
